### PR TITLE
Improve handling of concurrent load requests with server expanded layers

### PR DESCRIPTION
**Pull request by chuckdumont**
This pull request resolves a performance issue with server expanded layers.  The issue arises when the responses to overlapping loader requests arrive out of order (i.e. in different order than they were issued).  Because of the cumulative nature of the module exclude list that is maintained by the loader extension JavaScript, out of order responses can have unresolved dependencies (for dependent modules that were excluded by virtue of being dependent modules of previously issued, but still pending, request).  The loader doesn't know that these unsatisfied dependencies will be satisfied when the pending requests complete because it didn't request them by name.  They are dependent modules of the modules that were requested.  As a result of the undefined modules, additional, often cascading, requests are sent to  load the missing dependencies until the responses for the original pending requests arrive.

This pull request resolves that issue by intercepting and then queuing the define() function calls for out-of-order responses until the responses for all previously issued requests have arrived, and then playing back the define() function calls in request order.

### CODE REVIEW

**Comment by chuckdumont** [comment]
@ddumont

**Comment by ddumontatibm** [comment]
LGTM, but is there some kind of high-level documentation for the module/layer building process?   I havent actually been in the code recently so I'm starting to get a bit lost with all the changes to it over the past year or so.

Also... is there something we might want to ask dojo to take into their loader to help us here?  or are you happy with overriding define?

**Comment by chuckdumont** [comment]
You got me to thinking of an alternative approach that doesn't require overriding define.  Will update the PR soon.

**Comment by chuckdumont** [comment]
@ddumont, pull request updated to avoid overriding define().  Added more javadoc to document what is going on. See javadoc for DojoHttpTransport.beginModules().

**Comment by chuckdumont** [comment]
@ddumont please review

File: jaggr-core/WebContent/dojo/_loaderExt.js [hunk 1]
```diff
@@ -53,13 +53,20 @@ var depmap = {},
...
+    excludes = [],
+    
+    // Array of pending loads.   The values are {mids:xxx, cb:xxx} objects where
+    // mids is the same module list use to make the key and cb is a function that when called
+    // will define the modules in the layer.
+    pendingLoads = [],
+    
     // Query arg from window.location
-    windowArgs = parseQueryArgs(window.location.search) || {};
+    windowArgs = parseQueryArgs(window.location.search) || {},
```
Comment (ddumontatibm) [inline jaggr-core/WebContent/dojo/_loaderExt.js]: good catch

File: jaggr-core/src/test/java/com/ibm/jaggr/core/impl/layer/LayerBuilderTest.java [hunk 2]
```diff
@@ -126,11 +126,23 @@ public String answer() throws Throwable {
 							Assert.assertNull(arg);
 							return "]";
 						case BEGIN_MODULES:
-							Assert.assertNull(arg);
-							return "(";
+						{
+							StringBuffer sb = new StringBuffer("(`");
```
Comment (ddumontatibm) [inline jaggr-core/src/test/java/com/ibm/jaggr/core/impl/layer/LayerBuilderTest.java]: Can you just use google's 
```java
Joiner.on(';').join((Set<String>)arg);
```

**Comment by ddumontatibm** [comment]
otherwise lgtm

**Comment by chuckdumont** [comment]
Updated to use Joiner.

**Pull request merged**